Description: Add Ubuntu id into search queries and modify user agent to get
 credit.
Author: Chad Miller <chad.miller@canonical.com>
Author: Nathan Pratta Teodosio <nathan.teodosio@canonical.com>
Forwarded: not-needed

--- a/components/search_engines/prepopulated_engines.json
+++ b/components/search_engines/prepopulated_engines.json
@@ -48,7 +48,7 @@
       "name": "\u767e\u5ea6",
       "keyword": "baidu.com",
       "favicon_url": "https://www.baidu.com/favicon.ico",
-      "search_url": "https://www.baidu.com/#ie={inputEncoding}&wd={searchTerms}",
+      "search_url": "https://www.baidu.com/#ie={inputEncoding}&tn=ubuntuu_cb&wd={searchTerms}",
       "alternate_urls": [
         "https://www.baidu.com/s?ie={inputEncoding}&wd={searchTerms}",
         "https://www.baidu.com/s?ie={inputEncoding}&word={searchTerms}",
@@ -95,7 +95,7 @@
       "name": "DuckDuckGo",
       "keyword": "duckduckgo.com",
       "favicon_url": "https://duckduckgo.com/favicon.ico",
-      "search_url": "https://duckduckgo.com/?q={searchTerms}",
+      "search_url": "https://duckduckgo.com/?q={searchTerms}&t=canonical",
       "suggest_url": "https://duckduckgo.com/ac/?q={searchTerms}&type=list",
       "new_tab_url": "https://duckduckgo.com/chrome_newtab",
       "type": "SEARCH_ENGINE_DUCKDUCKGO",
--- a/components/search_engines/template_url.cc
+++ b/components/search_engines/template_url.cc
@@ -995,6 +995,9 @@ std::string TemplateURLRef::HandleReplac
 
   std::string url = parsed_url_;
 
+  bool is_google_search = false;
+  bool ubuntu_credit_for_search = false;
+
   // replacements_ is ordered in ascending order, as such we need to iterate
   // from the back.
   for (const Replacement& replacement : base::Reversed(replacements_)) {
@@ -1117,6 +1120,7 @@ std::string TemplateURLRef::HandleReplac
 
       case GOOGLE_BASE_URL:
         DCHECK(!replacement.is_post_param);
+        is_google_search = true;
         HandleReplacement(std::string(), search_terms_data.GoogleBaseURLValue(),
                           replacement, &url);
         break;
@@ -1243,7 +1247,8 @@ std::string TemplateURLRef::HandleReplac
 
       case GOOGLE_SEARCH_CLIENT: {
         DCHECK(!replacement.is_post_param);
-        std::string client = search_terms_data.GetSearchClient();
+        ubuntu_credit_for_search = true;
+        std::string client = "ubuntu-chr";
         if (!client.empty())
           HandleReplacement("client", client, replacement, &url);
         break;
@@ -1392,6 +1397,13 @@ std::string TemplateURLRef::HandleReplac
     }
   }
 
+  /* Google search template from prefs might not have a client token. */
+  if (is_google_search && !ubuntu_credit_for_search) {
+    if (url.find_first_of('?', 8) != std::string::npos) {  // perhaps no GET params
+      url.append("&client=ubuntu-chr");
+    }
+  }
+
   if (!post_params_.empty())
     EncodeFormData(post_params_, post_content);
 
--- a/content/common/user_agent.cc
+++ b/content/common/user_agent.cc
@@ -30,7 +30,7 @@
 namespace {
 
 const char kFrozenUserAgentTemplate[] =
-    "Mozilla/5.0 (%s) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/%s.0.0.0 "
+    "Mozilla/5.0 (X11; Ubuntu; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/%s.0.0.0 "
 #if BUILDFLAG(IS_ANDROID)
     "%s"
 #endif
@@ -64,7 +64,7 @@
 #elif BUILDFLAG(IS_FUCHSIA)
   return "Fuchsia";
 #elif BUILDFLAG(IS_LINUX)
-  return "X11; Linux x86_64";
+  return "X11; Ubuntu; Linux x86_64";
 #else
 #error Unsupported platform
 #endif
@@ -295,7 +295,7 @@
   device_compat = mobile ? "Mobile " : "";
 #endif
   std::string user_agent =
-      base::StringPrintf(kFrozenUserAgentTemplate, GetUnifiedPlatform().c_str(),
+      base::StringPrintf(kFrozenUserAgentTemplate,
                          major_version.c_str()
 #if BUILDFLAG(IS_ANDROID)
                              ,
@@ -382,9 +382,9 @@
   // compatible with Safari, we hope!!
   std::string user_agent;
   base::StringAppendF(&user_agent,
-                      "Mozilla/5.0 (%s) AppleWebKit/537.36 (KHTML, like Gecko) "
+                      "Mozilla/5.0 (X11; Ubuntu; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) "
                       "%s Safari/537.36",
-                      os_info.c_str(), product.c_str());
+                      product.c_str());
   return user_agent;
 }
 
